target/linux/mediatek/image/mt7623.mk

```diff
@@ -1,12 +1,41 @@
 KERNEL_LOADADDR := 0x80008000
 
+ifneq ($(CONFIG_BANANA_PI_BOOT_PARTSIZE),)
+BOOTFS_BLOCK_SIZE := 1024
+BOOTFS_BLOCKS := $(shell echo $$(($(CONFIG_BANANA_PI_BOOT_PARTSIZE)*1024*1024/$(BOOTFS_BLOCK_SIZE))))
+endif
+
+define Build/banana-pi-sdcard
+	rm -f $@.boot
+	mkfs.fat -C $@.boot $(BOOTFS_BLOCKS)
+
+	./gen_banana_pi_img.sh emmc $@.emmc \
+		$(STAGING_DIR_IMAGE)/$(UBOOT_TARGET)-preloader.bin
+
+	mcopy -i $@.boot $(STAGING_DIR_IMAGE)/$(UBOOT_TARGET)-uEnv.txt ::uEnv.txt
+	mcopy -i $@.boot $(IMAGE_KERNEL) ::uImage
+	mcopy -i $@.boot $@.emmc ::eMMCboot.bin
+	./gen_banana_pi_img.sh sd $@ \
+		$(STAGING_DIR_IMAGE)/$(UBOOT_TARGET)-preloader.bin \
+		$(STAGING_DIR_IMAGE)/$(UBOOT_TARGET)-u-boot.bin \
+		$@.boot \
+		$(IMAGE_ROOTFS) \
+		$(CONFIG_BANANA_PI_BOOT_PARTSIZE) \
+		$(CONFIG_TARGET_ROOTFS_PARTSIZE)
+endef
+
 define Device/bpi_bananapi-r2
   DEVICE_VENDOR := Bpi
   DEVICE_MODEL := Banana Pi R2
   DEVICE_DTS := mt7623n-bananapi-bpi-r2
   SUPPORTED_DEVICES := bananapi,bpi-r2
   DEVICE_PACKAGES := kmod-fs-vfat kmod-nls-cp437 kmod-nls-iso8859-1 kmod-mmc \
 	mkf2fs e2fsprogs kmod-usb-ohci kmod-usb2 kmod-usb3 kmod-ata-ahci-mtk
+  UBOOT_TARGET := mt7623n_bpir2
+  IMAGES := img.gz
+  IMAGE/img.gz += banana-pi-sdcard | gzip | append-metadata
+  KERNEL := kernel-bin | fit none $$(DTS_DIR)/$$(DEVICE_DTS).dtb
+  KERNEL_INITRAMFS :=
 endef
```

Comment:
Right now I see only **squashfs** images are generated. Can this line **FILESYSTEMS += ext4** be added in the device definition section to be able to generate ext4 images for this target also ?

Reply:
I thought there was pushback against that. So I dropped it and we can add it later perhaps after we've done the basic support. And once the sysupgrade support is tested.

Reply:
I have cloned the OpenWRT trunk, applied all your changes from this pull request and compiled a new ext4 image. I have wrote the image on the SD card but when trying to browse on the Internet every page loads very slow. It takes more than 2 minutes and some of them time out. Ping works fine (I left all settings at the defaults)
I saw this problem appeared a while ago after kernel 5.4.32 was released and something has been broken since then because on my OpenWRT fork using kernel 5.4.31 everything runs fine. If you can test please let me know if you experience the same issue.

Reply:
Hmm, interesting. That FILESYSTEMS is not listed in DEFAULT_DEVICE_VARS. Either it's hidden somewhere or defining it per-device will create problems ...

Reply:
Most of my testing has been the wan device, copying an image to the device for sysupgrade. It hasn't seemed slow. Can you do a packet capture and see what's happening? Are there dropped packets, bad checksums?

https://github.com/lexa2/openwrt/commit/5ec02dbe03958e39db292ca4ac7d94bfab9aea72 comes to mind.

Reply:
Also, are you seeing this with IPv6 or Legacy IP or both?

Reply:
Yes with both protocols IPv4 or IPv6 is the same. I can barely get 5 Mbps in both directions. I have a 500Mbps/250Mbps connection. I did not apply that patch lexa2@5ec02db in my fork it works. No dropped packets or any other interface errors on **wan** or **lan** interfaces only the speed is throttled.

Reply:
What is the network topology? Is this routing *through* the Banana Pi? Just routing, or also NAT.

Either way I'd be surprised if this PR makes the difference; can you test the OpenWrt build (with its convoluted install instructions) from current master?

Reply:
Right now my Banana-Pi R2 router is behind another TP-Link router which has the public IP on it. The WAN port (DHCP) is connected into the TP-link router and the LAN ports are bridged. I'm using masquerading/NAT to access Internet.
I compiled the image yesterday from OpenWRT master and applied the commits from your pull request and the situation it's the same.

Reply:
I've plugged a client into the a LAN port of mine (which was still configured for NAT as it's a clean installation) and run iperf to a machine on the house network, so it's mostly like your setup. I'm currently connected to an audio conference call through it, which is working perfectly and iperf reports 660Mb/s TCP at the same time.
```
[  3] local 192.168.1.228 port 36468 connected with 90.155.92.213 port 5001
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-30.0 sec  2.31 GBytes   660 Mbits/sec
```
>  I compiled the image yesterday from OpenWRT master and applied the commits from your pull request and the situation it's the same.

The idea was to test it *without* my commits, so that we know it isn't introduced by them and we can have this conversation elsewhere and actually merge this PR :)

Reply:
> Update: the problem appears to be on the LAN side. I ran wget directly on the box and download several iso images and the speed is fine only when trying from the LAN things go worse.

Do you have multiple devices in LAN ports? Can you unplug all but one and does it make a difference? Not that it should, but that's the main difference I can think of between your test and mine...

Reply:
No I only have my desktop connected behind it. I also did an iperf test and I got around 900 Mbps from my PC to BananaPi R2 but in the opposite direction I barely get few kilobytes
root@OpenWrt:/# iperf3 -c 192.168.1.220
Connecting to host 192.168.1.220, port 5201
[  5] local 192.168.1.1 port 60766 connected to 192.168.1.220 port 5201
[ ID] Interval                 Transfer        Bitrate               Retr
[  5]   0.00-10.00  sec  45.6 KBytes  37.4 Kbits/sec   18             sender
[  5]   0.00-10.00  sec  15.7 KBytes  12.8 Kbits/sec                  receiver

Reply:
I get about 800Mb/s sending from the client, 940Mb/s sending from the BPi.

Reply:
Can you capture with tcpdump on the PC and share the cap file?

Reply:
I have compiled again today a new clean squashfs image for BPI-R2 target with your commits included and the situation is the same. I have saved a small tcpdump pcap file which captured the iperf test from the BananaPi R2 router to my local PC. I have uploaded the pcap file here https://pcap.honeynet.org.my/v1/submission/downloadpcap.php?id=12089

Reply:
> have compiled again today a new clean squashfs image for BPI-R2 target with your commits included and the situation is the same.

Do you mean *without* my commits?

Reply:
No I mean I have applied all your commits to the OpenWRT master repository which I cloned it locally then compiled the image. Without your commits I think building the image won't work since the current OpenWRT master does not have u-boot and preloader packages added.

Reply:
Right, but you can use the ones you have, and follow the existing (convoluted) install instructions on the OpenWrt hardware page for the R2.

I don't think my patches are the cause of this.

Reply:
I have tried that already to apply my patches to a newer version of the OpenWRT repo (If I remember correctly around 2 weeks ago and the issue is there). I agree also that your patches are not the cause for this problem. My working OpenWRT fork is from 2 months ago using kernel 5.4.31. During this time newer kernel versions were pushed are also some new patches were added here openwrt/target/linux/mediatek/patches-5.4/.

Reply:
Should be relatively simple to bisect that and find which update broke it.

Reply:
Let's discuss that elsewhere and close the discussion here.

Reply:
@cioby23 Did you find the Ethernet problem? I think I'm seeing something similar on the U7623 now.

Reply:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable-rc.git/commit/?h=linux-5.4.y&id=1ff0732cf8055131050fa0c26c49429d6e725505 went into 5.4.34, which seems to be around the right time frame to suspect it. But it doesn't seem to be that.

@blogic any thoughts?

Reply:
I dropped all of 0999-hnat.patch, 1000-eth-gdm-config-backport.patch and 1001-eth-mtk-disable-TBDDONE-support.patch too, which came in after the version you said was working. Those don't help either.


```
root@OpenWrt:/# iperf3 -c i7.infradead.org
Connecting to host i7.infradead.org, port 5201
[  5] local 2001:8b0:10b:1::c2a port 49498 connected to 2001:8b0:10b:1:21e:67ff:fecb:7a92 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   130 KBytes  1.06 Mbits/sec   26   1.39 KBytes       
[  5]   1.00-2.00   sec  0.00 Bytes  0.00 bits/sec    1   1.39 KBytes       
[  5]   2.00-3.00   sec  0.00 Bytes  0.00 bits/sec    1   1.39 KBytes       
[  5]   3.00-4.00   sec  0.00 Bytes  0.00 bits/sec    1   1.39 KBytes       
[  5]   4.00-5.00   sec  0.00 Bytes  0.00 bits/sec    0   1.39 KBytes       
[  5]   5.00-6.00   sec  0.00 Bytes  0.00 bits/sec    0   1.39 KBytes       
[  5]   6.00-7.00   sec  0.00 Bytes  0.00 bits/sec    0   1.39 KBytes       
[  5]   7.00-8.00   sec  0.00 Bytes  0.00 bits/sec    1   1.39 KBytes       
[  5]   8.00-9.00   sec  0.00 Bytes  0.00 bits/sec    0   1.39 KBytes       
[  5]   9.00-10.00  sec  0.00 Bytes  0.00 bits/sec    0   1.39 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   130 KBytes   106 Kbits/sec   30             sender
[  5]   0.00-10.00  sec  80.9 KBytes  66.2 Kbits/sec                  receiver

iperf Done.
root@OpenWrt:/# iperf3 -c i7.infradead.org -R
Connecting to host i7.infradead.org, port 5201
Reverse mode, remote host i7.infradead.org is sending
[  5] local 2001:8b0:10b:1::c2a port 49502 connected to 2001:8b0:10b:1:21e:67ff:fecb:7a92 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec   111 MBytes   927 Mbits/sec                  
[  5]   1.00-2.00   sec   110 MBytes   924 Mbits/sec                  
[  5]   2.00-3.00   sec   107 MBytes   899 Mbits/sec                  
[  5]   3.00-4.00   sec   107 MBytes   902 Mbits/sec                  
[  5]   4.00-5.00   sec   107 MBytes   899 Mbits/sec                  
[  5]   5.00-6.00   sec   107 MBytes   900 Mbits/sec                  
[  5]   6.00-7.00   sec   107 MBytes   900 Mbits/sec                  
[  5]   7.00-8.00   sec   107 MBytes   899 Mbits/sec                  
[  5]   8.00-9.00   sec   108 MBytes   902 Mbits/sec                  
[  5]   9.00-10.00  sec   108 MBytes   903 Mbits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  1.06 GBytes   907 Mbits/sec   54             sender
[  5]   0.00-10.00  sec  1.05 GBytes   905 Mbits/sec                  receiver

iperf Done.
```
```

Reply:
Should be fixed in #3230 at least for Legacy IP. For IPv6 it's back to at least 2/3 of line rate, often more. In the 5.8.0-rc6 kernel I get 9xxMb/s in both directions for both IPv6 and Legacy IP with the patch from https://github.com/frank-w/BPI-R2-4.14/commit/ed15a1f84cfe2d731f20658b3c2dab3f7b3bc648